File: ms-bridge-redis.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
ms_bridge_redis.py
Consume pyEfi Megasquirt realtime data from Redis PubSub and expose a HUDIY-friendly WS/HTTP API.

WebSocket:
  ws://0.0.0.0:8765/v1/stream

HTTP:
  GET /v1/channels  -> {"fields": { ... }}
  GET /v1/snapshot  -> last frame
  GET /v1/health    -> status

Run examples:
  # UNIX socket (pyEfi default)
  python3 ms_bridge_redis.py --redis-sock /var/run/redis/redis.sock

  # TCP
  python3 ms_bridge_redis.py --redis-host 127.0.0.1 --redis-port 6379

  # Custom channel and publish rate
  python3 ms_bridge_redis.py --channel ms2:realtime --hz 20
"""

import argparse, asyncio, json, time, signal, contextlib
from dataclasses import dataclass, field
from typing import Dict, Any, Optional, Tuple

from aiohttp import web, WSMsgType
import redis.asyncio as redis  # async redis-py

# ---------------------- Config ----------------------

@dataclass
class BridgeConfig:
    hz: float = 20.0
    ws_host: str = "0.0.0.0"
    ws_port: int = 8765
    channel: str = "ms2:realtime"
    redis_sock: Optional[str] = None
    redis_host: Optional[str] = None
    redis_port: int = 6379

    fields: Dict[str, Dict[str, str]] = field(default_factory=lambda: {
        "rpm":     {"unit": "rpm"},
        "map_kpa": {"unit": "kPa"},
        "tps_pct": {"unit": "%"},
        "clt_f":   {"unit": "°F"},
        "afr":     {"unit": ""},
        "batt_v":  {"unit": "V"},
        "mat_f":   {"unit": "°F"},
    })

# ---------------------- Normalization ----------------------

def _first(d: Dict[str, Any], *keys) -> Optional[Any]:
    for k in keys:
        if k in d and d[k] is not None:
            return d[k]
        # case-insensitive lookup
        for kk in d.keys():
            if kk.lower() == k.lower() and d[kk] is not None:
                return d[kk]
    return None

def _to_float(x) -> Optional[float]:
    try:
        return float(x)
    except Exception:
        return None

def _c_to_f(val_c: Optional[float]) -> Optional[float]:
    if val_c is None:
        return None
    return val_c * 9.0 / 5.0 + 32.0

def normalize_payload(src: Dict[str, Any]) -> Dict[str, Any]:
    """Map pyEfi payload into stable keys expected by the HUDIY gauges."""
    # rpm
    rpm = _to_float(_first(src, "rpm", "RPM", "engine.rpm"))
    # map (prefer kPa)
    map_kpa = _to_float(_first(src, "map_kpa", "map", "MAP", "engine.map"))
    # tps (%)
    tps = _to_float(_first(src, "tps_pct", "tps", "TPS"))
    # CLT + MAT (accept either F already or C -> convert)
    clt = _to_float(_first(src, "clt_f", "CLT_f", "Coolant_f", "coolant_f", "cltF"))
    if clt is None:
        clt_c = _to_float(_first(src, "clt", "Coolant", "CLT"))
        clt = _c_to_f(clt_c) if clt_c is not None else None

    mat = _to_float(_first(src, "mat_f", "IAT_f", "matF", "iat_f"))
    if mat is None:
        mat_c = _to_float(_first(src, "mat", "IAT"))
        mat = _c_to_f(mat_c) if mat_c is not None else None

    afr = _to_float(_first(src, "afr", "AFR", "lambda", "Lambda"))
    batt_v = _to_float(_first(src, "batt_v", "vBatt", "bat", "battery"))

    out = {}
    if rpm is not None:     out["rpm"] = int(round(rpm))
    if map_kpa is not None: out["map_kpa"] = round(map_kpa, 1)
    if tps is not None:     out["tps_pct"] = round(tps, 1)
    if clt is not None:     out["clt_f"] = round(clt, 1)
    if afr is not None:     out["afr"] = round(afr, 2)
    if batt_v is not None:  out["batt_v"] = round(batt_v, 2)
    if mat is not None:     out["mat_f"] = round(mat, 1)
    return out

def parse_pubsub_message(msg) -> Optional[Dict[str, Any]]:
    """Handle bytes/str payload, return dict or None."""
    if msg["type"] != "message":
        return None
    payload = msg.get("data")
    if payload is None:
        return None
    if isinstance(payload, (bytes, bytearray)):
        try:
            payload = payload.decode("utf-8", "replace")
        except Exception:
            return None
    if isinstance(payload, str):
        payload = payload.strip()
        try:
            return json.loads(payload)
        except Exception:
            # sometimes pyEfi may already give dict (rare), or non-json: ignore
            return None
    if isinstance(payload, dict):
        return payload
    return None

# ---------------------- Redis consumer ----------------------

class RedisReader:
    def __init__(self, cfg: BridgeConfig):
        self.cfg = cfg
        self._r: Optional[redis.Redis] = None
        self._pubsub = None
        self.last_data_ts: float = 0.0
        self.connected: bool = False

    async def connect(self):
        if self._r:
            await self.close()
        if self.cfg.redis_sock:
            self._r = redis.Redis(unix_socket_path=self.cfg.redis_sock)
        else:
            host = self.cfg.redis_host or "127.0.0.1"
            port = int(self.cfg.redis_port or 6379)
            self._r = redis.Redis(host=host, port=port)
        self._pubsub = self._r.pubsub()
        await self._pubsub.subscribe(self.cfg.channel)
        self.connected = True

    async def close(self):
        try:
            if self._pubsub:
                await self._pubsub.close()
        finally:
            self._pubsub = None
        try:
            if self._r:
                await self._r.close()
        finally:
            self._r = None
            self.connected = False

    async def next_message(self, timeout: float = 1.0) -> Optional[Dict[str, Any]]:
        """Wait for next message (async); returns normalized dict or None on timeout."""
        if not (self._r and self._pubsub):
            return None
        try:
            msg = await self._pubsub.get_message(ignore_subscribe_messages=True, timeout=timeout)
            if not msg:
                return None
            raw = parse_pubsub_message(msg)
            if not raw:
                return None
            self.last_data_ts = time.time()
            return normalize_payload(raw)
        except Exception:
            return None

# ---------------------- Bridge app (WS/HTTP) ----------------------

class BridgeApp:
    def __init__(self, cfg: BridgeConfig, reader: RedisReader):
        self.cfg = cfg
        self.reader = reader
        self.app = web.Application()
        self.app.add_routes([
            web.get("/v1/stream", self.ws_handler),
            web.get("/v1/channels", self.channels_handler),
            web.get("/v1/snapshot", self.snapshot_handler),
            web.get("/v1/health", self.health_handler),
        ])
        self.clients = set()
        self.last_frame: Optional[Dict[str, Any]] = None
        self._producer_task: Optional[asyncio.Task] = None

    async def start(self):
        self._producer_task = asyncio.create_task(self.producer_loop())

    async def channels_handler(self, request):
        return web.json_response({"fields": self.cfg.fields})

    async def snapshot_handler(self, request):
        return web.json_response(self.last_frame or {"type": "frame", "ts": time.time(), "data": {}})

    async def health_handler(self, request):
        age_ms = int(1000 * (time.time() - self.reader.last_data_ts)) if self.reader.last_data_ts else None
        return web.json_response({
            "ok": bool(self.reader.connected),
            "connected": self.reader.connected,
            "lastDataAgoMs": age_ms,
            "clients": len(self.clients),
            "channel": self.cfg.channel,
        })

    async def ws_handler(self, request):
        ws = web.WebSocketResponse()
        await ws.prepare(request)
        self.clients.add(ws)
        try:
            await ws.send_str(json.dumps({"type": "hello", "note": "pyEfi/Redis bridge"}))
            if self.last_frame:
                await ws.send_str(json.dumps(self.last_frame))
            async for msg in ws:
                if msg.type == WSMsgType.TEXT:
                    if msg.data.strip().lower() == "ping":
                        await ws.send_str("pong")
                elif msg.type == WSMsgType.ERROR:
                    break
        finally:
            self.clients.discard(ws)
        return ws

    async def producer_loop(self):
        period = 1.0 / max(self.cfg.hz, 1.0)
        backoff = 0.5
        while True:
            try:
                if not self.reader.connected:
                    await self.reader.connect()
                    backoff = 0.5

                # Try to pull a fresh message; if none, reuse last data
                data = await self.reader.next_message(timeout=period)
                if data is None and self.last_frame:
                    # keep broadcasting last known data at steady Hz
                    data = self.last_frame["data"]

                if data is not None:
                    frame = {"type": "frame", "ts": time.time(), "data": data}
                    self.last_frame = frame
                    if self.clients:
                        payload = json.dumps(frame, separators=(",", ":"))
                        await asyncio.gather(*[
                            c.send_str(payload) for c in list(self.clients) if not c.closed
                        ], return_exceptions=True)

                await asyncio.sleep(period)

            except Exception:
                # On any error, close and backoff reconnect
                await self.reader.close()
                await asyncio.sleep(backoff)
                backoff = min(backoff * 1.7, 8.0)

# ---------------------- Main ----------------------

def parse_args():
    ap = argparse.ArgumentParser()
    ap.add_argument("--hz", type=float, default=20.0, help="broadcast rate to WS clients")
    ap.add_argument("--wshost", default="0.0.0.0")
    ap.add_argument("--wsport", type=int, default=8765)
    ap.add_argument("--channel", default="ms2:realtime")

    grp = ap.add_mutually_exclusive_group()
    grp.add_argument("--redis-sock", help="UNIX socket path, e.g. /var/run/redis/redis.sock")
    grp.add_argument("--redis-host", help="Redis host (if not using UNIX socket)")
    ap.add_argument("--redis-port", type=int, default=6379)

    return ap.parse_args()

async def main_async():
    args = parse_args()
    cfg = BridgeConfig(
        hz=args.hz,
        ws_host=args.wshost, ws_port=args.wsport,
        channel=args.channel,
        redis_sock=args.redis_sock,
        redis_host=args.redis_host,
        redis_port=args.redis_port,
    )

    reader = RedisReader(cfg)
    app = BridgeApp(cfg, reader)

    runner = web.AppRunner(app.app)
    await runner.setup()
    site = web.TCPSite(runner, host=cfg.ws_host, port=cfg.ws_port)
    await site.start()
    await app.start()

    print(f"[ws] ws://{cfg.ws_host}:{cfg.ws_port}/v1/stream")
    print(f"[http] /v1/channels  /v1/snapshot  /v1/health")
    print(f"[redis] channel={cfg.channel} via " + (cfg.redis_sock or f"{cfg.redis_host or '127.0.0.1'}:{cfg.redis_port}"))
    print(f"[rate] {cfg.hz} Hz")

    stop = asyncio.Future()
    loop = asyncio.get_running_loop()
    for sig in (signal.SIGINT, signal.SIGTERM):
        loop.add_signal_handler(sig, lambda: (not stop.done()) and stop.set_result(True))
    await stop

def main():
    try:
        asyncio.run(main_async())
    except KeyboardInterrupt:
        pass

if __name__ == "__main__":
    main()

File: ms-bridge-redis.js
#!/usr/bin/env node
/**
 * ms-bridge-redis.js
 * Consume pyEfi Megasquirt realtime JSON from Redis Pub/Sub and expose a HUDIY-friendly WS/HTTP API.
 *
 * WebSocket:
 *   ws://0.0.0.0:8765/v1/stream
 *
 * HTTP:
 *   GET /v1/channels  -> {"fields": {...}}
 *   GET /v1/snapshot  -> last frame
 *   GET /v1/health    -> status (connected, lastDataAgoMs, clients, channel)
 *
 * Run examples:
 *   # UNIX socket (pyEfi default)
 *   node ms-bridge-redis.js --redis-sock=/var/run/redis/redis.sock
 *
 *   # TCP
 *   node ms-bridge-redis.js --redis-host=127.0.0.1 --redis-port=6379
 *
 *   # Custom channel and broadcast rate
 *   node ms-bridge-redis.js --channel=ms2:realtime --hz=20
 */

import { createServer } from 'http';
import { WebSocketServer } from 'ws';
import Redis from 'ioredis';

// ---------------- CLI / ENV ----------------
const args = Object.fromEntries(
  process.argv.slice(2).map(a => {
    const [k, v] = a.split('=');
    return [k.replace(/^--/, ''), v ?? true];
  })
);

const HZ = Number(args.hz || process.env.HZ || 20);
const WS_HOST = args.wshost || process.env.WS_HOST || '0.0.0.0';
const WS_PORT = Number(args.wsport || process.env.WS_PORT || 8765);
const CHANNEL = args.channel || process.env.REDIS_CHANNEL || 'ms2:realtime';

// Redis: prefer UNIX socket, else TCP
const REDIS_SOCK = args['redis-sock'] || process.env.REDIS_SOCK || '';
const REDIS_HOST = args['redis-host'] || process.env.REDIS_HOST || '127.0.0.1';
const REDIS_PORT = Number(args['redis-port'] || process.env.REDIS_PORT || 6379);

// ---------------- Normalization helpers ----------------
function firstKey(obj, variants) {
  for (const key of variants) {
    if (obj[key] != null) return obj[key];
    const found = Object.keys(obj).find(k => k.toLowerCase() === String(key).toLowerCase());
    if (found && obj[found] != null) return obj[found];
  }
  return undefined;
}
function toFloat(x) {
  if (x == null) return undefined;
  const n = Number(x);
  return Number.isFinite(n) ? n : undefined;
}
function cToF(c) {
  if (c == null) return undefined;
  return c * 9 / 5 + 32;
}

// Map pyEfi payload to stable HUDIY keys
function normalizePayload(src) {
  // rpm
  const rpm = toFloat(firstKey(src, ['rpm', 'RPM', 'engine.rpm']));
  // map (kPa preferred)
  const map_kpa = toFloat(firstKey(src, ['map_kpa', 'map', 'MAP', 'engine.map']));
  // tps in %
  const tps = toFloat(firstKey(src, ['tps_pct', 'tps', 'TPS']));

  // CLT and MAT (accept °F directly or convert from °C)
  let clt_f = toFloat(firstKey(src, ['clt_f', 'CLT_f', 'Coolant_f', 'coolant_f', 'cltF']));
  if (clt_f == null) {
    const clt_c = toFloat(firstKey(src, ['clt', 'Coolant', 'CLT']));
    if (clt_c != null) clt_f = cToF(clt_c);
  }

  let mat_f = toFloat(firstKey(src, ['mat_f', 'IAT_f', 'matF', 'iat_f']));
  if (mat_f == null) {
    const mat_c = toFloat(firstKey(src, ['mat', 'IAT']));
    if (mat_c != null) mat_f = cToF(mat_c);
  }

  const afr = toFloat(firstKey(src, ['afr', 'AFR', 'lambda', 'Lambda']));
  const batt_v = toFloat(firstKey(src, ['batt_v', 'vBatt', 'bat', 'battery']));

  const out = {};
  if (rpm != null) out.rpm = Math.round(rpm);
  if (map_kpa != null) out.map_kpa = Number(map_kpa.toFixed(1));
  if (tps != null) out.tps_pct = Number(tps.toFixed(1));
  if (clt_f != null) out.clt_f = Number(clt_f.toFixed(1));
  if (afr != null) out.afr = Number(afr.toFixed(2));
  if (batt_v != null) out.batt_v = Number(batt_v.toFixed(2));
  if (mat_f != null) out.mat_f = Number(mat_f.toFixed(1));
  return out;
}

function parseRedisMessage(msg) {
  // ioredis 'message' gives (channel, messageString)
  if (typeof msg !== 'string' && !(msg instanceof Buffer)) return null;
  let s = msg;
  if (msg instanceof Buffer) s = msg.toString('utf8');
  s = String(s).trim();
  try {
    const parsed = JSON.parse(s);
    if (parsed && typeof parsed === 'object') return parsed;
  } catch {
    // not JSON, ignore
  }
  return null;
}

// ---------------- Bridge state ----------------
const fields = {
  rpm:     { unit: 'rpm' },
  map_kpa: { unit: 'kPa' },
  tps_pct: { unit: '%' },
  clt_f:   { unit: '°F' },
  afr:     { unit: '' },
  batt_v:  { unit: 'V' },
  mat_f:   { unit: '°F' }
};

let lastFrame = null;
let lastDataTs = 0;
const clients = new Set();

// ---------------- Redis subscriber ----------------
function makeRedisSubscriber() {
  const opts = REDIS_SOCK
    ? { path: REDIS_SOCK }
    : { host: REDIS_HOST, port: REDIS_PORT };

  const sub = new Redis(opts);

  sub.on('connect', () => {
    console.log(`[redis] connecting ${REDIS_SOCK ? REDIS_SOCK : `${REDIS_HOST}:${REDIS_PORT}`}`);
  });
  sub.on('ready', async () => {
    console.log(`[redis] ready, subscribing "${CHANNEL}"`);
    try { await sub.subscribe(CHANNEL); }
    catch (e) { console.error('[redis] subscribe error:', e.message); }
  });
  sub.on('close', () => console.warn('[redis] connection closed'));
  sub.on('reconnecting', (t) => console.warn(`[redis] reconnecting in ${t}ms`));
  sub.on('error', (e) => console.error('[redis] error:', e.message));

  sub.on('message', (_channel, message) => {
    const raw = parseRedisMessage(message);
    if (!raw) return;
    const data = normalizePayload(raw);
    if (!data || Object.keys(data).length === 0) return;

    lastDataTs = Date.now() / 1000;
    lastFrame = { type: 'frame', ts: lastDataTs, data };
  });

  return sub;
}

const subscriber = makeRedisSubscriber();

// ---------------- WS + HTTP server ----------------
const server = createServer((req, res) => {
  try {
    if (req.url === '/v1/channels') {
      res.writeHead(200, { 'content-type': 'application/json' });
      res.end(JSON.stringify({ fields }));
      return;
    }
    if (req.url === '/v1/snapshot') {
      res.writeHead(200, { 'content-type': 'application/json' });
      res.end(JSON.stringify(lastFrame || { type: 'frame', ts: Date.now() / 1000, data: {} }));
      return;
    }
    if (req.url === '/v1/health') {
      const age = lastDataTs ? (Date.now() / 1000 - lastDataTs) : null;
      const ok = subscriber.status === 'ready' || subscriber.status === 'connecting' || subscriber.status === 'reconnecting' ? true : false;
      res.writeHead(200, { 'content-type': 'application/json' });
      res.end(JSON.stringify({
        ok,
        connected: subscriber.status === 'ready',
        lastDataAgoMs: age != null ? Math.round(age * 1000) : null,
        clients: clients.size,
        channel: CHANNEL
      }));
      return;
    }

    res.writeHead(404);
    res.end();
  } catch (e) {
    res.writeHead(500);
    res.end(JSON.stringify({ error: e.message }));
  }
});

const wss = new WebSocketServer({ server, path: '/v1/stream' });
wss.on('connection', (ws) => {
  clients.add(ws);
  try {
    ws.send(JSON.stringify({ type: 'hello', note: 'pyEfi/Redis bridge' }));
    if (lastFrame) ws.send(JSON.stringify(lastFrame));
  } catch {}
  ws.on('close', () => clients.delete(ws));
});

// steady broadcast loop @ HZ (re-broadcasts last good data)
const periodMs = Math.max(5, Math.round(1000 / Math.max(1, HZ)));
setInterval(() => {
  if (!lastFrame || clients.size === 0) return;
  const payload = JSON.stringify(lastFrame);
  for (const ws of clients) {
    if (ws.readyState === ws.OPEN) {
      try { ws.send(payload); } catch {}
    }
  }
}, periodMs);

server.listen(WS_PORT, WS_HOST, () => {
  console.log(`[ws] ws://${WS_HOST}:${WS_PORT}/v1/stream`);
  console.log(`[http] /v1/channels  /v1/snapshot  /v1/health`);
  console.log(`[redis] channel=${CHANNEL} via ${REDIS_SOCK ? REDIS_SOCK : `${REDIS_HOST}:${REDIS_PORT}`}`);
  console.log(`[rate] ${HZ} Hz`);
});

// graceful shutdown
function shutdown() {
  try { subscriber.quit(); } catch {}
  try { server.close(() => process.exit(0)); } catch { process.exit(0); }
}
process.on('SIGINT', shutdown);
process.on('SIGTERM', shutdown);
